### DISCUSSION TITLE
Build tools and waiting on flux to be 'ready'?

### DISCUSSION THREAD

**Discussion by landbaychrisburrell** [fluxcd/flux2#4431] [Q&A]
Hi

I've read various discussions here and blogs on testing our current set-up with flux. I've got most things working, except the 'waiting on flux' part. I spin up a KIND cluster, using DinD on Gitlab, and can then bootstrap flux.

However, here are my issues
* the `bootstrap` command doesn't wait for the whole cluster to be set-up (so far as I can tell). It seems to return as soon as flux is bootstrapped (makes sense)
* I then run `reconcile `command with a timeout of 10 minutes but also returns pretty quickly. 
* When running flux get all -A subsequently, I can I have two "unknown" readyness statuses (Reconcilitation in progress) as well as a large number of dependent items that are not ready.

Is there a single command i can run to ensure I can wait until Flux has finished everything?

Also, Flux CLI doesn't seem to return response codes? Would that be a nice addition - e.g. if reconcile in fact returns "all ready" would be nice to get a different response code to "some still reconciling".

Perhaps I'm doing something wrong? Here's the commands that i run from above:
```
flux -v
export GITLAB_TOKEN=<>
flux bootstrap gitlab --insecure-skip-tls-verify --owner=myowner --repository=my-repo --branch=master --path=clusters/mycluster--deploy-token-auth --read-write-key
flux get all -A --insecure-skip-tls-verify
export REC_RESPONSE=`flux reconcile kustomization flux-system -n flux-system --with-source --timeout 10m  --insecure-skip-tls-verify`
echo "Response code from response is ${REC_RESPONSE}"
flux get all -A --insecure-skip-tls-verify
```

Essentially, I just want to make sure Flux has finished before I start running assertions against the cluster. What ways can you recommend for achieving this?

**Comment by kingdonb** [comment]
Check out the example provided in flux2-kustomize-helm example `e2e.yaml`:

https://github.com/fluxcd/flux2-kustomize-helm-example/blob/0ad4dfb50bbba3f3ab2b661aa82c931c2111fb2f/.github/workflows/e2e.yaml#L34-L37

You can set the `spec.wait` on your Flux kustomizations to be sure they have health checks enabled on the resources that they apply, then run something like:

```
kubectl -n flux-system wait kustomization/infra-controllers --for=condition=ready --timeout=5m
kubectl -n flux-system wait kustomization/apps --for=condition=ready --timeout=5m
```

The ready condition will not be reached unless the pods are all firing and ready, happy resources all become available. This command will not return before the Flux resources are available (without returning an error)

(So, in your script, you might also need to set `-e`

```
set -e The set -e option instructs bash to immediately exit if any command [1] has a non-zero exit status.
```

https://gist.github.com/vncsna/64825d5609c146e80de8b1fd623011ca)

Hope this helps!

**Reply by landbaychrisburrell** [reply]
Hi

Thanks for the answer - that's really helpful - will I need to add the 'wait' on each and every child resource - e.g. if I have a Kustomization pointing to another Kustomization, I assume it's not sufficient to just put at the top level? e.g. I can't put this on the top level cluster kustomization?

Cheers
Chris

**Reply by stefanprodan** [reply]
You need to wait for the last child

**Reply by stefanprodan** [reply]
In our e2e repo, we wait for `kustomization/apps` because this one depends on infra-config that depends on infra. Waiting for it, means all its dependencies are completed.
